### PR TITLE
[MRG+1]fix bug in estimate_stc

### PR DESCRIPTION
**Pull request by makkostya**
To simulate an stc from label the function `estimate_stc` uses the intersection between label vertices and source space. However, if label contains only one or few vertices (for example only the center of mass of a region) the intersection can be empty. This possibility is not currently taken into account which may result in  outputting empty stcs.

I added the code which returns a source even if the intersection is empty. For a moment I just select the source from source space whose index is the nearest (from right) to the label vertices.

### CODE REVIEW

**Comment by makkostya** [comment]
cc @sdeslauriers @ngayraud @ikojcic

**Comment by agramfort** [comment]
can you add a test @makkostya ?

thanks a lot

**Comment by agramfort** [comment]
Ready to switch to MRG?

File: mne/simulation/source.py [hunk 1]
```diff
@@ -267,6 +267,13 @@ def simulate_stc(src, labels, stc_data, tmin, tstep, value_fun=None,
         hemi_ind = hemi_to_ind[label.hemi]
         src_sel = np.intersect1d(src[hemi_ind]['vertno'],
                                  label.vertices)
+        if len(src_sel) == 0:
+            idx = np.searchsorted(src[hemi_ind]['vertno'],
+                                  label.vertices)
+            src_len = len(src[hemi_ind]['vertno']) - 1
+            idx[idx > src_len] = src_len
+            src_sel = np.unique(src[hemi_ind]['vertno'][idx])
+
```
Comment (massich) [inline mne/simulation/source.py]: when I see `src_sel` I read: selected sources, or sorces that are {selected, active}. This makes me think of a list of some sort (in this case represented by a numpy array).

More over, the block of code in this PR shows that this is the case. 'cos it reads: `if the list of selected sources is empty do something`. What breaks all this logic is that then you reuse this very same variable to store a scalar, perform a selection use it back as a recipient of the computation and then there is no else statement.

I would recommend you to create a function that would make everything much more readable (choose `foo`, `bar` and `foobar` approperly):

```py
src_sel = _select_foobar(foo=src[hemi_ind]['vertno'], bar=label.vertices)
```

Then when you write the function I would use:
```py
def _select_foobar(foo, bar):
    intersection = np.intersect1d(src[hemi_ind]['vertno'], labels.vertices)
    if intersection.size > 0:  # This is how np recomends to check for not empty
         return intersection
    else:
         return something else
```

**Comment by massich** [comment]
@makkostya 	you should update whatsnew

File: mne/simulation/source.py (hunk 1 above)
Comment (massich) [reply mne/simulation/source.py]: Still, the snipped its complicated to read. 

`src_sel` it ends up being `[54]` which is the same as `src[hemi_ind]['vertno'][0]`

I'm confused.

Comment (makkostya) [reply mne/simulation/source.py]: > src_sel it ends up being [54] which is the same as src[hemi_ind]['vertno'][0]

that's what it should be in test.

My label contains a single vertex with index 0:
`i = np.where(fwd['src'][0]['inuse'] == 0)[0][0]`

This vertex is not in the source space. So the closest vertex index from the source space to label vertex is  `[54] `

File: mne/simulation/source.py [hunk 2]
```diff
@@ -267,6 +267,13 @@ def simulate_stc(src, labels, stc_data, tmin, tstep, value_fun=None,
         hemi_ind = hemi_to_ind[label.hemi]
         src_sel = np.intersect1d(src[hemi_ind]['vertno'],
                                  label.vertices)
+        if len(src_sel) == 0:
+            idx = np.searchsorted(src[hemi_ind]['vertno'],
+                                  label.vertices)
+            src_len = len(src[hemi_ind]['vertno']) - 1
+            idx[idx > src_len] = src_len
+            src_sel = np.unique(src[hemi_ind]['vertno'][idx])
```
Comment (agramfort) [inline mne/simulation/source.py]: @makkostya by writing this do you assume that if you get an idx that is close to label.vertices that it is physically close to the label on the cortex? if so this is likely broken as you cannot guarantee this. Either you need to grow the label of I would suggest to indeed write a meaningful error message in an exception

following an IRL discussion with @massich

File: mne/simulation/source.py (hunk 1 above)
Comment (massich) [reply mne/simulation/source.py]: factoring in a function and creating toy examples to fully tested in a isolated enviroment would provide better understanding

File: mne/simulation/source.py (hunk 2 above)
Comment (makkostya) [reply mne/simulation/source.py]: @agramfort And what if select the vertex from the source space which is the closest in terms of Euclidean norm?

Comment (massich) [reply mne/simulation/source.py]: To me that would make sense. 
To sum up the problem, if I understand properly, we have 2 sets of identifiers (which they all have position somewhere else). Let's assume that those sets are number (id) and their position is (id, 0, 0). Then 1 is colser to 2 than to 3 and so on.
(for sake of abstraction I'll call src[hemi_ind]['vertno']-> foo, and  label.vertices->bar)

case A: there is the intersection of the identifiers
```py
    foo=[3, 6, 9]
    bar=[0, 1, 2, 3, 4, 5, 6]  
    result = my_func(foo, bar) # [3, 6]
```

case B: there is no intersection, return set(the closest foo id of the bar element)
```py
    foo=[3, 6, 9]
    bar=[0, 1, 2, 3, 4, 5, 6]  
    result = my_func(foo, bar) # [3, 3, 3, 3, 3, 6, 5]
    resut = set(result) # [3, 6]
```

Then would happen in someting like:
```py
my_func(foo=[3, 6, 9], bar=[0, 1, 2, 3, 4, 5, 7, 8]) # [3]
my_func(foo=[3, 6, 9], bar=[0, 1, 2, 4, 5, 7, 8]) # [3, 6, 9]
```

Question, do we care about the size of the resulting set? is it better `[3]` than `[3, 6, 9]`?

Maybe we should always take the set of the closest element (with a threshold if we consider that the hull of two sets might no intersect)

**Comment by agramfort** [comment]
That could work. You don’t want to raise an error to suggest to grow the label ?

**Comment by makkostya** [comment]
> You don’t want to raise an error to suggest to grow the label ?

I want the code to work without user doing extra codding. For example if user wants to simulate a dipole in the center of a label, he shouldn't have to pay attention if the output of `label.center_of_mass()` is or isn't in the source space. For instance, `mne.label.select_sources(subject, label, location='center', extent=0)` does not work properly.

**Comment by agramfort** [comment]
if you take the center of mass restricted to the label vertices i would
work no?

>

**Comment by makkostya** [comment]
It would return the vertex from the set of label vertices. But this vertex does not have to be in the source space, which is usually a sub-sampling of cortical mesh.

**Comment by agramfort** [comment]
true select_sources is not aware of the source space. Then I would say
find the closest
vertices in the source space.

**Comment by agramfort** [comment]
@massich merge if you're happy and when CIs are green

**Comment by massich** [comment]
Just to make sure that we are all in the same page are we find with [this comment](https://github.com/mne-tools/mne-python/pull/6516#discussion_r300363745)? (replicate below)



---

Let's assume that those sets are number (id) and their position is (id, 0, 0). Then 1 is colser to 2 than to 3 and so on.
(for sake of abstraction I'll call `src[hemi_ind]['vertno']` -> `foo`, and  `label.vertices`->`bar`)

case A: there is the intersection of the identifiers
```py
    foo=[3, 6, 9]
    bar=[0, 1, 2, 3, 4, 5, 6]  
    result = my_func(foo, bar) # [3, 6]
```

case B: there is no intersection, return set(the closest foo id of the bar element)
```py
    foo=[3, 6, 9]
    bar=[0, 1, 2, 3, 4, 5, 6]  
    result = my_func(foo, bar) # [3, 3, 3, 3, 3, 6, 5]
    resut = set(result) # [3, 6]
```

Then would happen in someting like:
```py
my_func(foo=[3, 6, 9], bar=[0, 1, 2, 3, 4, 5, 7, 8]) # [3]
my_func(foo=[3, 6, 9], bar=[0, 1, 2, 4, 5, 7, 8]) # [3, 6, 9]
```

Question, do we care about the size of the resulting set? is it better `[3]` than `[3, 6, 9]`?

Maybe we should always take the set of the closest element (with a threshold if we consider that the hull of two sets might no intersect)

**Comment by makkostya** [comment]
> Maybe we should always take the set of the closest element (with a threshold if we consider that the hull of two sets might no intersect)

I would use closest sources only in the case when the intersection of label and source space is empty as a precaution to not have empty stc.

btw, do you know what caused the CI to fail?

**Comment by larsoner** [comment]
> btw, do you know what caused the CI to fail?

Looks like a spurious download error on Azure

**Comment by makkostya** [comment]
So,  is it ready to merge?

**Comment by agramfort** [comment]
Ok for me

@massich @larsoner please merge if happy

**Comment by larsoner** [comment]
Thanks @makkostya

**Pull request merged**